### PR TITLE
Native VPC support

### LINKED ISSUE #6673: Native VPC support - Simple setup
Native VPC support via `vpc: true` config:

```yaml
service: vpc-test

vpc: true
```

It should create:

1. A VPC
1. Public and private subnets (1 public, 1 private)
1. InternetGateway to enable internet access for public subnets
1. A security group for the Lambda functions.

Furthermore it should export the VPC setup so that it can be used across stacks.

Shoutouts to @alexdebrie who came up with the spec.

### PR DESCRIPTION
**Pull request by pmuens**
Closes #6673
Refs #6674

## What did you implement:

Adds native vpc support via dead-simple `vpc: true` config.

**NOTE:** This PR also includes ways to change the CIDR setup as shown below.

## How did you implement it:

Add a new `package:compileVpc` lifecycle and implemented the plugin to hook into this. The plugin in turn will create the CloudFormation resources according to #6673.

## How can we verify it:

Deploy the following service with the `--verbose` flag to see the outputs (`serverless deploy --verbose`).

### Simple setup

```yaml
service: test-${self:custom.idx}

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

vpc: true

custom:
  idx: 0

functions:
  hello:
    handler: handler.handler
```

### Extended setup

```yaml
service: test-${self:custom.idx}

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

vpc:
  cidrBlock: 11.0.0.0/16
  subnets:
    private:
      cidrBlock: 11.0.1.0/24
    public:
      cidrBlock: 11.0.2.0/24

custom:
  idx: 0

functions:
  hello:
    handler: handler.handler
```

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

- [x] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Ensure introduced changes match Prettier formatting.  
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

**Comment by pmuens** [comment]
@alexdebrie @medikoo this PR is now ready for a final review. Would love to get your feedback on it!

**Comment by jplock** [comment]
If its of any help, I have a plugin that does a lot of this including provisioning a bastion host and a NAT instance instead of a NAT Gateway at https://github.com/smoketurner/serverless-vpc-plugin

File: lib/plugins/aws/package/compile/vpc/index.test.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    it('should have hooks', () => expect(awsCompileVpc.hooks).to.be.not.empty);
+
+    it('should set the provider variable to be an instanceof AwsProvider', () =>
+      expect(awsCompileVpc.provider).to.be.instanceof(AwsProvider));
+
+    describe('"package:compileVpc" promise chain', () => {
+      it('should run the promise chain in order', () => {
+        return awsCompileVpc.hooks['package:compileVpc']().then(() => {
+          expect(compileVpcStub.calledOnce).to.be.equal(true);
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/vpc/index.test.js]: To be honest, I'm not convinced whether such tests are really valid

As what is tested here, is not the outcome that really counts: the expected VPC configuration that lands in CF template, but instead we test, whether given internal methods were called and how many times.
It seems that what we're really guarding here, is that this specific implementation is written line by line written as it's currently, and it's not the thing we should really care about I believe

It raises maintenance issues, as whenever we will refactor the implementation, so e.g.  it's leaner (without changing the outcome), those tests will have to be refactored as well - that makes future improvements to the framework more cumbersome to do.

Ideally tests should guard the outcome (desired effect), and not internals of implementation. They should allow to do painless refactors, where we can overhaul the implementation to something very different, and via test run confirm that all works as is it worked before.

Imo right approach is to treat the internals of a plugin as black box. Test the expected result, and via coverage confirm that we covered all cases envisioned in implementation.

I would prepare just one test file for VPC plugin, and would rely on `@serverless/test/run-serverless` util, which ensures you run complete instance, so eventual side effects introduced by other plugins are also in play.

Currently it'll require to create you a dummy service folder (with simplest possible `serverless.yml` that has `vpc: true`) (as side note, we can improve that util so it takes config via argument (instead of `cwd`)).

Having that you may prepare a `serverless` instance ready for testing as:

```javascript
runServerless({
  cwd: vpcServicePath,
  pluginPathsWhitelist: [vpcPluginPath],
  hookNamesWhitelist: ['package:compileVpc'],
})).then(serverless => {
  const cfTemplate = serverless.service.provider.compiledCloudFormationTemplate;
  // Confirm that cfTemplate has VPC configuration we're after
});
```

What do you think about that?

File: lib/plugins/aws/package/compile/vpc/index.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            .then(this.compileVpcGatewayAttachment)
+            .then(this.compileRouteTables)
+            .then(this.compileSubnets)
+            .then(this.compileNatGateways)
+            .then(this.compileRoutes)
+            .then(this.compileSubnetRouteTableAssociations)
+            .then(this.compileLambdaSecurityGroup)
+            .then(this.updateFunctionConfigs)
+            .then(this.compileCloudFormationOutputs)
+            .then(this.compileSfeOutputs);
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/vpc/index.js]: As I checked, all those functions are sync. Why do we make them _async_?

Imo trying to handle _sync_ stuff, as if it's _async_, unnecessarily increases the complexity of the setup, and also confuses programmers trying to read through internals.

What is the point of that approach?

Technically it could be written as simply as:

```javascript
compileVpc(serverless);
compileEip(serverless);
compileInternetGateway(serverless);
...
```

File: lib/plugins/aws/package/compile/vpc/index.test.js (hunk 1 above)
Comment (pmuens) [reply lib/plugins/aws/package/compile/vpc/index.test.js]: Totally agree with that. I guess the root cause for such tests were the chasing of test coverage increases in the past.

So the idea is that I completely remove the unit tests for this plugin? I'm not sure about that because for large implementations (like this) it can be really tough to track down the actual root cause for the test failure.

But yes, let me prepare such test with our new `runServerless` capability (really like the API 👍).

Comment (medikoo) [reply lib/plugins/aws/package/compile/vpc/index.test.js]: > So the idea is that I completely remove the unit tests for this plugin?  

Yes, I would ditch all tests that check whether we called given function and how many times 
and just run tests that confirm on generated template.

> I'm not sure about that because for large implementations (like this) it can be really tough to track down the actual root cause for the test failure.

I wonder whether that can be a real issue. Technically implementation should be quite self explanatory, so if we see that that there's an issue in generated resource X or that it's missing, it shouldn't be a big trouble to track down the internal function that's responsible for that.

At least testing that depends on specificity of internals, puts on us hand-cuffs, that whenever we shuffle internals, we also should update tests, and that doesn't say good of tests imo.

On unit tests level, I would only leave tests that confirm some non trivial algorithms (e.g. fahrenheit o celsius conversion, deciding which year is leap year). Still in many cases such functions are rather meant to be excluded into utils, and over there such unit tests naturally make big sense.

So I guess good approach would be:
- Test outcome of a plugin as if it's a black box (tests should not check on it's implementation internals, how many and what functions are created internally and which order invoked)
- Exclude any internal non trivial logic units into utils (if it's questionable to expose them globally we may create `utils.js` in plugin directory), which should be accompanied with unit tests that test this util individually. (Still I don't think there's a case of such utils in context of this plugin ?)

Comment (pmuens) [reply lib/plugins/aws/package/compile/vpc/index.test.js]: I agree with the sentiment to focus on integration test rather than starting with unit tests. However very often there are nuanced edge cases which would require a lot of different integration test setups. Such edge cases are easier tested with unit tests and then the integration test is used to only cover the "happy" / simple path.

I'd suggest to default to integration tests only but if there are some non-trivial edge cases we should mix in unit tests for such scenarios.

Comment (medikoo) [reply lib/plugins/aws/package/compile/vpc/index.test.js]: Sure, if for some reason just via natural plugin evaluation, it's hard to test some scenarios, then indeed we probably should explore alternatives, and maybe introduce unit tests for some isolated part.

Still even in such case, I think it'll be good to ensure that tests do not depend too much on  structure of internals.

Do we have such cases here?

Comment (pmuens) [reply lib/plugins/aws/package/compile/vpc/index.test.js]: > Still even in such case, I think it'll be good to ensure that tests do not depend too much on structure of internals.

Makes sense 👍

> Do we have such cases here?

Not that I'm aware of 🤔

Comment (pmuens) [reply lib/plugins/aws/package/compile/vpc/index.test.js]: Just removed the unit tests and replaced them with an integration test.

File: lib/plugins/aws/package/compile/vpc/index.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            }
+            if (vpc.subnets && vpc.subnets.public) {
+              const { cidrBlock } = vpc.subnets.public;
+              subnetConfigs.public = {
+                cidrBlock,
+              };
+            }
+          }
+
+          this.compileVpc.call(this, vpcConfig);
```
[the remark this answers is not in the document]
Comment (pmuens) [reply lib/plugins/aws/package/compile/vpc/index.js]: 🤦‍♂

**Comment by pmuens** [comment]
Thanks for the review @medikoo :+1:

Just updated the PR and addressed all your review comments! 👍

> It really makes a notable improvement, and new tests setup, looks as decent upgrade over what we're practicing. Through them it's also now easier to learn, what's the real outcome of a plugin (as a whole)

Yep, after we got them up- and running yesterday it really felt like home and it was a real relief to delete all the separate unit tests.

**Comment by medikoo** [comment]
Packaging integration tests now fail for some reason

**Comment by alexdebrie** [comment]
Really excited for this, @pmuens 🙌.

One question -- Do you think it's possible our defaults could change in breaking ways in the future? And if so, should we have an optional `version` property or something, so that users that deployed with older defaults won't break their existing deploys?

Example:

**Basic usage:**

```yml
service: vpc-infra

vpc: true
```

**With version property:**

```yml
service: vpc-infra

vpc:
  version: 1
```

**With version property and other properties:**

This would use all the defaults for version 1, but the `cidrBlock` property is given a custom value:

```yml
service: vpc-infra

vpc:
  version: 1
  cidrBlock: 11.0.0.0/16 
```

File: lib/plugins/aws/package/compile/vpc/lib/subnets.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+const { SUBNET_TYPES, SUBNET_CONFIG } = require('./utils/constants');
+const { flatMap } = require('./utils/functions');
+const getSubnetTemplate = require('./templates/subnet');
+
+function compileSubnets(subnetConfigs) {
+  const { Resources } = this.serverless.service.provider.compiledCloudFormationTemplate;
+  const types = [SUBNET_TYPES.PRIVATE, SUBNET_TYPES.PUBLIC];
+  // NOTE: previously we loaded the supported regions from `./utils/constants`
+  // we keep the `flatMap` code intact to allow multi-region azs later on
+  const availabilityZones = flatMap(region => [`${region}a`], [this.options.region]);
```
Comment (alexdebrie) [inline lib/plugins/aws/package/compile/vpc/lib/subnets.js]: I'm confused on the `availabilityZones` part -- does this only use the `a` availability zone for the region? The docs say that the default config creates subnets in both `a` and `b` AZs for the given region -- how is that accomplished? The test outputs only show `a` from what I can tell.

Also, FWIW, it might be useful to use the [`GetAZs`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getavailabilityzones.html) intrinsic function in CloudFormation for this.

File: docs/providers/aws/guide/vpc.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+## VPC Resources
+
+Using the `vpc: true` configuration will instruct the Serverless Framework to setup the following resources in your AWS account:
+
+- 1 VPC
+- 2 Subnets (1 private, 1 public) in the `a` and `b` availability zones of your deployment region
+- 1 InternetGateway to enable internet access for public subnets
+- 1 SecurityGroup for your Lambda functions
```
Comment (alexdebrie) [inline docs/providers/aws/guide/vpc.md]: Do these need to be updated? Looks like we're only configuring in a single AZ (comment elsewhere), and we're configuring additional resources -- an EIP and a NAT Gateway.

**Comment by jplock** [comment]
@alexdebrie Thanks for the kind words about my plugin!  I've been a long-time user and fan of this framework.  My thoughts were based on the Skyliner service which no longer exists, but really centers around the default routes of each subnet:

* public subnet -> internet gateway
* app subnet -> NAT Gateway or NAT Instance
* db subnet -> no default route

If the lambda functions need to access the Internet, and are within a VPC, the default route of the subnet needs to be a NAT Gateway or a NAT Instance (an Internet Gateway won't work).  Even to reach AWS services, you need to either use a NAT Gateway, NAT Instance or have VPC Endpoints available.

File: lib/plugins/aws/package/compile/vpc/lib/natGateways.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+const { SUBNET_TYPES } = require('./utils/constants');
```
[the remark this answers is not in the document]
Comment (jplock) [reply lib/plugins/aws/package/compile/vpc/lib/natGateways.js]: Lambda functions can't reach the Internet through an Internet Gateway, only NAT Instances or NAT Gateways.

File: lib/plugins/aws/package/compile/vpc/lib/subnets.js (hunk 4 above)
Comment (jplock) [reply lib/plugins/aws/package/compile/vpc/lib/subnets.js]: Similar to what my plugin does, but would it be possible to make a call to the AWS API to get the available AZs in a region?  Sometimes they can be marked as unavailable so I'm not sure we should rely on looking for `a` or `b` in the naming convention.

**Comment by alexdebrie** [comment]
@jplock As I understand it, that link suggests putting Lambdas in a _private_ subnet. I'm wondering if we can put the Lambdas in a _public_ subnet. That way, they get internet access via an Internet Gateway, but they can also access private resources via the route table that allows intra-VPC traffic.

I don't think we'd need to worry about inbound traffic hitting the Lambda through the public subnet, as I'm pretty sure Lambdas are only accessible via a single, special VPC that AWS manages.

That said, I have enough doubt from reading that link that I'm not 100% sure :) Need to do some testing.

**Comment by alexdebrie** [comment]
Also, @jplock , could you email me at alex@serverless.com? Got a question for you 😄

**Comment by pmuens** [comment]
Thanks for everyone involved in this (@jplock @alexdebrie @medikoo @dschep) :+1:

We've decided to put this on hold for the time being and explore other ways to introduce this functionality. Given that I'll close this PR for now.

Everyone who's reading this. Feel free to continue the discussion / provide feedback. It would be awesome to hear from you if / how you would use such feature...
